### PR TITLE
Fix notify state destruction and inflight states tracking

### PR DESCRIPTION
**Pull request by tanmayv25**
For the cases where gRPC doesn't deliver notification, we have moved the notify_state_ ~delete to the context  destructor~ to the unique_ptr.
This should fix the valgrind failures on CI.

Fixing the tracking of inflight states to erase the states as and when they complete or run into an error. This should fix the L0_perf_analyzer under heavy streaming load case.

### CODE REVIEW

File: src/grpc/infer_handler.h [hunk 1]
```diff
@@ -627,22 +627,29 @@ class InferHandlerState {
...
+
     void SetCompressionLevel(grpc_compression_level compression_level)
     {
       ctx_->set_compression_level(compression_level);
     }
 
     void GrpcContextAsyncNotifyWhenDone(InferHandlerStateType* state)
     {
-      InferHandlerStateType* wrapped_state =
+      notify_state_ =
```
Comment (rmccorm4) [inline src/grpc/infer_handler.h:650]: If there is only one notify state per context, can this just be a `unique_ptr` or something to auto clean up when context is destructed?

Comment (tanmayv25) [reply src/grpc/infer_handler.h:650]: That's a good idea!

**Comment by matthewkotila** [comment]
Thanks for fixing this bug, but I'll defer to others for the review, since I am not familiar with this code.

**Review by kthui — APPROVED** [review]

File: src/grpc/infer_handler.h [hunk 2]
```diff
@@ -1274,7 +1282,6 @@ InferHandler<
         state->context_->SetReceivedNotification(true);
         LOG_VERBOSE(1) << "Received notification for " << Name() << ", "
                        << state->unique_id_;
-        delete state_wrapper;
```
[the remark this answers is not in the document]
Comment (tanmayv25) [reply src/grpc/infer_handler.h:1277]: The state_wrapper ptr(the current notify_stat_ ptr), is now a unique_ptr within the context object. This means we would not need to explicitly delete the object.

**Review by dyastremsky — APPROVED** [review]

File: src/grpc/infer_handler.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (dyastremsky) [reply src/grpc/infer_handler.h:1277]: I see, thanks for explaining! It does look like the old state_wrapper exists in the code (unchanged). Just an FYI.

**Pull request merged**